### PR TITLE
KAFKA-3616: Make kafka producers/consumers injectable for KafkaStreams

### PR DESCRIPTION
**Pull request by kawamuray**
Ticket: https://issues.apache.org/jira/browse/KAFKA-3616

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 1]
```diff
@@ -98,16 +98,21 @@
     private final UUID processId;
 
     public KafkaStreams(TopologyBuilder builder, Properties props) {
-        this(builder, new StreamsConfig(props));
+        this(builder, new StreamsConfig(props), new DefaultKafkaClientSupplier());
+    }
+
+    public KafkaStreams(TopologyBuilder builder, StreamsConfig config) {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Good catch!

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 2]
```diff
@@ -98,16 +98,21 @@
...
+        this(builder, config, new DefaultKafkaClientSupplier());
     }
 
     /**
      * Construct the stream instance.
      *
      * @param builder The processor topology builder specifying the computational logic
      * @param config The stream configs
+     * @param clientSupplier The kafka clients supplier which provides clients for this KafkaStream
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: nit: "which provide underlying producer and consumer clients for this {@code KafkaStream} instance".

**Comment by guozhangwang** [comment]
@kawamuray Could you rebase?

A couple of general comments:

1) Move DefaultClientSupplier into streams/processor/internals as it does not need to be public interface.
2) Could you also create a MockClientSupplier which uses the MockProducer and MockConsumer and add a unit test to validate that the clients are generated as expected?

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java (hunk 2 above)
Comment (kawamuray) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: ack. maybe `@link` though? Fixed by:    8f45081

**Comment by kawamuray** [comment]
@guozhangwang Thanks for reviewing, rebased and addressed your comments with following up unit test fixes.

For your two comments:

> 1) Move DefaultClientSupplier into streams/processor/internals as it does not need to be public interface.

Actually I was thinking below as the practical use since I'm expecting that in most cases only producer or consumer is the injection target. What do you think of?

``` java
KafkaStreams streams = new KafkaStreams(topology, config, new DefaultKafkaClientSupplier() {
    @Override
    public Producer<byte[], byte[]> getProducer(Map<String, Object> config) {
        return new MyKafkaProducer<>(...);
    }
});
```

> 2) Could you also create a MockClientSupplier which uses the MockProducer and MockConsumer and add a unit test to validate that the clients are generated as expected?

Ack. Addressed by: c2a5a5a and 8b13ced.

**Comment by guozhangwang** [comment]
LGTM overall, Jenkins failure are not related (filed separate JIRA).

About the DefaultClientSupplier: in general I'd prefer to keep things private until we see common requirement for them to be public, since vice versa cannot be easily done due to backward incompatibility. Thoughts?

**Comment by kawamuray** [comment]
> About the DefaultClientSupplier: in general I'd prefer to keep things private until we see common requirement for them to be public, since vice versa cannot be easily done due to backward incompatibility. Thoughts?

Okay, as all methods of DefaultKafkaClientSupplier are fairly easy to re-implement by users, let me follow your advice ATM :+1:

**Comment by guozhangwang** [comment]
Merged to trunk and cherry-picked to 0.10.0.

**Pull request closed without merging**